File: agent-python/agent/agent.py
import asyncio
import logging
from typing import List, Literal, Optional

from . import messages
from .grenade_info import GrenadeInfo
from .map import Map
from .player_info import FirearmKind, Item, ItemKind, PlayerInfo
from .position import Position
from .safe_zone import SafeZone
from .supply import Supply, SupplyKind
from .websocket_client import WebsocketClient

MedicineKind = Literal[
    "BANDAGE",
    "FIRST_AID",
]


class Agent:
    def __init__(self, token: str, loop_interval: float):
        self._token = token
        self._loop_interval = loop_interval

        self._all_player_info: Optional[List[PlayerInfo]] = None
        self._map: Optional[Map] = None
        self._supplies: Optional[List[Supply]] = None
        self._safe_zone: Optional[SafeZone] = None
        self._grenade_info: Optional[List[GrenadeInfo]] = None
        self._self_id: Optional[int] = None
        self._ticks: Optional[int] = None

        self._ws_client = WebsocketClient()
        self._loop_task: Optional[asyncio.Task] = None

        self._ws_client.on_message = self._on_message

    def __str__(self) -> str:
        return f"Agent{{token: {self._token}}}"

    def __repr__(self) -> str:
        return str(self)

    @property
    def all_player_info(self) -> Optional[List[PlayerInfo]]:
        return self._all_player_info

    @property
    def map(self) -> Optional[Map]:
        return self._map

    @property
    def supplies(self) -> Optional[List[Supply]]:
        return self._supplies

    @property
    def safe_zone(self) -> Optional[SafeZone]:
        return self._safe_zone

    @property
    def self_id(self) -> Optional[int]:
        return self._self_id

    @property
    def token(self) -> str:
        return self._token

    @property
    def ticks(self) -> Optional[int]:
        return self._ticks

    @property
    def grenade_info(self) -> Optional[List[GrenadeInfo]]:
        return self._grenade_info

    async def connect(self, server: str):
        await self._ws_client.connect(server)
        self._loop_task = asyncio.create_task(self._loop())

    async def disconnect(self):
        if self._loop_task is not None:
            self._loop_task.cancel()
        await self._ws_client.disconnect()

    def is_connected(self) -> bool:
        return self._ws_client.is_connected()

    def is_game_ready(self) -> bool:
        return (
            self._all_player_info is not None
            and self._map is not None
            and self._supplies is not None
            and self._safe_zone is not None
            and self._self_id is not None
            and self._ticks is not None
            and self._grenade_info is not None
        )

    async def abandon(self, supply: SupplyKind, count: int):
        logging.debug("%s.abandon(%s, %d)", self, supply, count)
        await self._ws_client.send(
            messages.PerformAbandonMessage(
                token=self._token,
                numb=count,
                target_supply=supply,
            )
        )

    async def pick_up(self, supply: SupplyKind, count: int):
        logging.debug("%s.pick_up(%s, %d)", self, supply, count)
        await self._ws_client.send(
            messages.PerformPickUpMessage(
                token=self._token, target_supply=supply, num=count
            )
        )

    async def switch_firearm(self, firearm: FirearmKind):
        logging.debug("%s.switch_firearm(%s)", self, firearm)
        await self._ws_client.send(
            messages.PerformSwitchArmMessage(
                token=self._token,
                target_firearm=firearm,
            )
        )

    async def use_medicine(self, medicine: MedicineKind):
        logging.debug("%s.use_medicine(%s)", self, medicine)
        await self._ws_client.send(
            messages.PerformUseMedicineMessage(
                token=self._token,
                medicine_name=medicine,
            )
        )

    async def use_grenade(self, position: Position[float]):
        logging.debug("%s.use_grenade(%s)", self, position)
        await self._ws_client.send(
            messages.PerformUseGrenadeMessage(
                token=self._token,
                target_position=messages.Position(x=position.x, y=position.y),
            )
        )

    async def move(self, position: Position[float]):
        logging.debug("%s.move(%s)", self, position)
        await self._ws_client.send(
            messages.PerformMoveMessage(
                token=self._token,
                destination=messages.Position(x=position.x, y=position.y),
            )
        )

    async def stop(self):
        logging.debug("%s.stop()", self)
        await self._ws_client.send(
            messages.PerformStopMessage(
                token=self._token,
            )
        )

    async def attack(self, position: Position[float]):
        logging.debug("%s.attack(%s)", self, position)
        await self._ws_client.send(
            messages.PerformAttackMessage(
                token=self._token,
                target_position=messages.Position(x=position.x, y=position.y),
            )
        )

    async def choose_origin(self, position: Position[float]):
        logging.debug("%s.choose_origin(%s)", self, position)
        await self._ws_client.send(
            messages.ChooseOriginMessage(
                token=self._token,
                origin_position=messages.Position(x=position.x, y=position.y),
            )
        )

    async def _loop(self):
        while True:
            try:
                await asyncio.sleep(self._loop_interval)

                if not self._ws_client.is_connected():
                    continue

                await self._ws_client.send(
                    messages.GetPlayerInfoMessage(
                        token=self._token,
                    )
                )

            except Exception as e:
                logging.error(f"{self} encountered an error in loop: {e}")

    def _on_message(self, message: messages.Message):
        try:
            msg_dict = message.msg
            msg_type = msg_dict["messageType"]

            if msg_type == "ERROR":
                logging.error(f"{self} got error from server: {msg_dict['message']}")

            elif msg_type == "PLAYERS_INFO":
                self._ticks = msg_dict["elapsedTicks"]
                self._all_player_info = [
                    PlayerInfo(
                        id=data["playerId"],
                        armor=data["armor"],
                        current_armor_health=data["current_armor_health"],
                        health=data["health"],
                        speed=data["speed"],
                        firearm=data["firearm"]["name"],
                        firearms_pool=[
                            weapon_data["name"] for weapon_data in data["firearms_pool"]
                        ],
                        range=data["firearm"]["distance"],
                        position=Position(
                            x=data["position"]["x"], y=data["position"]["y"]
                        ),
                        inventory=[
                            Item(kind=item["name"], count=item["num"])
                            for item in data["inventory"]
                        ],
                    )
                    for data in msg_dict["players"]
                ]

            elif msg_type == "MAP":
                self._map = Map(
                    length=msg_dict["length"],
                    obstacles=[
                        Position(
                            x=wall["wallPositions"]["x"], y=wall["wallPositions"]["y"]
                        )
                        for wall in msg_dict["walls"]
                    ],
                )

            elif msg_type == "SUPPLIES":
                self._supplies = [
                    Supply(
                        kind=supply["name"],
                        position=Position(
                            x=supply["position"]["x"], y=supply["position"]["y"]
                        ),
                        count=supply["numb"],
                    )
                    for supply in msg_dict["supplies"]
                ]

            elif msg_type == "SAFE_ZONE":
                self._safe_zone = SafeZone(
                    center=Position(
                        x=msg_dict["center"]["x"], y=msg_dict["center"]["y"]
                    ),
                    radius=msg_dict["radius"],
                )

            elif msg_type == "PLAYER_ID":
                self._self_id = msg_dict["playerId"]

            elif msg_type == "GRENADES":
                self._grenade_info = [
                    GrenadeInfo(
                        throwTick=grenade["throwTick"],
                        evaluatedPosition=Position(
                            x=grenade["evaluatedPosition"]["x"],
                            y=grenade["evaluatedPosition"]["y"],
                        ),
                    )
                    for grenade in msg_dict["grenades"]
                ]

        except Exception as e:
            logging.error(f"{self} failed to handle message: {e}")

File: agent-cpp/src/agent/agent.cc
#include "agent.h"

#include <fmt/format.h>
#include <hv/Event.h>
#include <hv/EventLoop.h>
#include <hv/WebSocketClient.h>
#include <hv/hloop.h>
#include <spdlog/spdlog.h>

#include <string>
#include <string_view>

#include "agent/grenade_info.h"
#include "agent/map.h"
#include "agent/message.h"
#include "agent/player_info.h"
#include "agent/position.h"
#include "agent/supply.h"
#include "message.h"

namespace thuai7_agent {

Agent::Agent(std::string_view token, hv::EventLoopPtr const& event_loop,
             int loop_interval)
    : event_loop_(event_loop), token_(token) {
  loop_timer_id_ =
      event_loop_->setInterval(loop_interval, [this](hv::TimerID) { Loop(); });

  ws_client_ = std::make_unique<hv::WebSocketClient>(event_loop);

  reconn_setting_t reconn_setting;
  reconn_setting.delay_policy = 0;  // Fixed delay.

  ws_client_->setReconnect(&reconn_setting);

  ws_client_->onmessage = [this](std::string const& msg) {
    OnMessage(Message(msg));
  };
}

Agent::~Agent() { event_loop_->killTimer(loop_timer_id_); }

void Agent::Connect(std::string_view server_address) {
  ws_client_->open(server_address.data());
  // std::this_thread::sleep_for(std::chrono::seconds(5));
}

auto Agent::IsConnected() const -> bool { return ws_client_->isConnected(); }

void Agent::Disconnect() { ws_client_->close(); }

auto Agent::IsGameReady() const -> bool {
  return all_player_info_.has_value() && map_.has_value() &&
         supplies_.has_value() && safe_zone_.has_value() &&
         self_id_.has_value() && ticks_.has_value() &&
         grenade_info_.has_value();
}

void Agent::Abandon(SupplyKind target_supply, int count) {
  spdlog::debug("{}.Abandon({}, {})", *this, target_supply, count);
  ws_client_->send(PerformAbandonMessage(count, token_, target_supply).json());
}

void Agent::PickUp(SupplyKind target_supply, int count) {
  spdlog::debug("{}.PickUp({}, {})", *this, target_supply, count);
  ws_client_->send(PerformPickUpMessage(token_, target_supply, count).json());
}

void Agent::SwitchFirearm(FirearmKind target_firearm) {
  spdlog::debug("{}.SwitchFirearm({})", *this, target_firearm);
  ws_client_->send(PerformSwitchArmMessage(token_, target_firearm).json());
}

void Agent::UseMedicine(MedicineKind target_medicine) {
  spdlog::debug("{}.UseMedicine({})", *this, target_medicine);
  ws_client_->send(PerformUseMedicineMessage(token_, target_medicine).json());
}

void Agent::UseGrenade(Position<float> const& position) {
  spdlog::debug("{}.UseGrenade({})", *this, position);
  ws_client_->send(PerformUseGrenadeMessage(token_, position).json());
}

void Agent::Move(Position<float> const& position) {
  spdlog::debug("{}.Move({})", *this, position);
  ws_client_->send(PerformMoveMessage(token_, position).json());
}

void Agent::Stop() {
  spdlog::debug("{}.Stop()", *this);
  ws_client_->send(PerformStopMessage(token_).json());
}

void Agent::Attack(Position<float> const& position) {
  spdlog::debug("{}.Attack({})", *this, position);
  ws_client_->send(PerformAttackMessage(token_, position).json());
}

void Agent::ChooseOrigin(Position<float> const& position) {
  spdlog::debug("{}.ChooseOrigin({})", *this, position);
  ws_client_->send(ChooseOriginMessage(token_, position).json());
}

void Agent::Loop() {
  try {
    if (!IsConnected()) {
      return;
    }

    ws_client_->send(GetPlayerInfoMessage(token_).json());
  } catch (std::exception const& e) {
    spdlog::error("{} encountered an error in loop: {}", *this, e.what());
  }
}

void Agent::OnMessage(Message const& message) {
  auto msg_dict = message.msg;
  try {
    auto msg_type = msg_dict["messageType"].get<std::string>();

    if (msg_type == "ERROR") {
      spdlog::error("{} got an error from server: {}", *this,
                    msg_dict["message"].get<std::string>());
    } else if (msg_type == "PLAYERS_INFO") {
      ticks_ = msg_dict["elapsedTicks"].get<int>();
      all_player_info_ = std::vector<PlayerInfo>();
      for (auto const& data : msg_dict["players"]) {
        auto player_id = data["playerId"].get<int>();
        auto armor = data["armor"].get<ArmorKind>();
        auto current_armor_health = data["current_armor_health"].get<float>();
        auto health = data["health"].get<int>();
        auto speed = data["speed"].get<float>();
        auto firearm = data["firearm"]["name"].get<FirearmKind>();
        std::vector<FirearmKind> firearms_pool;
        for (auto const& msg_firearm : data["firearms_pool"]) {
          firearms_pool.emplace_back(msg_firearm["name"].get<FirearmKind>());
        }
        auto range = data["firearm"]["distance"].get<float>();
        Position<float> position{data["position"]["x"].get<float>(),
                                 data["position"]["y"].get<float>()};
        std::vector<Item> inventory;
        for (auto const& msg_item : data["inventory"]) {
          inventory.emplace_back(Item{msg_item["name"].get<ItemKind>(),
                                      msg_item["num"].get<int>()});
        }
        all_player_info_->emplace_back(
            PlayerInfo{player_id, armor, current_armor_health, health, speed,
                       firearm, firearms_pool, range, position, inventory});
      }
    } else if (msg_type == "MAP") {
      auto length = msg_dict["length"].get<int>();
      std::vector<Position<int>> walls;
      for (auto const& msg_wall : msg_dict["walls"]) {
        walls.emplace_back(
            Position<int>{msg_wall["wallPositions"]["x"].get<int>(),
                          msg_wall["wallPositions"]["y"].get<int>()});
      }
      map_ = Map{length, walls};
    } else if (msg_type == "SUPPLIES") {
      supplies_ = std::vector<Supply>();
      for (auto const& msg_supply : msg_dict["supplies"]) {
        auto kind = msg_supply["name"].get<SupplyKind>();
        Position<float> position{msg_supply["position"]["x"].get<float>(),
                                 msg_supply["position"]["y"].get<float>()};
        auto count = msg_supply["numb"].get<int>();

        supplies_->emplace_back(Supply{kind, count, position});
      }
    } else if (msg_type == "SAFE_ZONE") {
      Position<float> center{msg_dict["center"]["x"].get<float>(),
                             msg_dict["center"]["y"].get<float>()};
      auto radius = msg_dict["radius"].get<float>();
      safe_zone_ = SafeZone{center, radius};
    } else if (msg_type == "PLAYER_ID") {
      self_id_ = msg_dict["playerId"].get<int>();
    } else if (msg_type == "GRENADES") {
      grenade_info_ = std::vector<GrenadeInfo>();
      for (auto const& msg_grenade : msg_dict["grenades"]) {
        auto throwTick = msg_grenade["throwTick"].get<int>();
        Position<float> position{
            msg_grenade["evaluatedPosition"]["x"].get<float>(),
            msg_grenade["evaluatedPosition"]["y"].get<float>()};

        grenade_info_->emplace_back(GrenadeInfo{throwTick, position});
      }
    }

  } catch (std::exception const& e) {
    spdlog::error("{} failed to handle a message: {}", *this, e.what());
  }
}

auto format_as(Agent const& object) -> std::string {
  return fmt::format("Agent{{token: {}}}", object.token());
}

}  // namespace thuai7_agent
